src/main/scala/mesosphere/marathon/core/storage/store/impl/cache/LazyCachingPersistenceStore.scala

```diff
@@ -140,22 +170,94 @@ class LazyCachingPersistenceStore[K, Category, Serialized](
+    toRemove: Int = versionedValueCacheConfig.purgeCount,
+    pRemoval: Double = versionedValueCacheConfig.pRemove)(f: () => Boolean): Unit =
+    while (f()) {
+      // randomly GC the versions
+      val counter = new AtomicInteger
+      versionedValueCache.retain { (k, v) =>
+        val x = scala.util.Random.nextDouble()
+        x > pRemoval || counter.incrementAndGet() > toRemove
       }
     }
+
+  protected def updateCachedVersions[V](
+    storageId: K,
+    version: OffsetDateTime,
+    category: Category,
+    v: V): Unit = {
+
+    maybePurgeCachedVersions() { () => versionedValueCache.size > versionedValueCacheConfig.maxEntries }
+    versionedValueCache.put((storageId, version), Some(v))
+    val cached = versionCache.getOrElse((category, storageId), Nil) // linter:ignore UndesirableTypeInference
+    versionCache.put((category, storageId), (version +: cached).distinct)
   }
 
-  override def versions[Id, V](id: Id)(implicit ir: IdResolver[Id, V, Category, K]): Source[OffsetDateTime, NotUsed] =
-    store.versions(id)
+  override def versions[Id, V](id: Id)(implicit ir: IdResolver[Id, V, Category, K]): Source[OffsetDateTime, NotUsed] = {
+    val category = ir.category
+    val storageId = ir.toStorageId(id, None)
+    val versionsFuture = lockManager.executeSequentially(category.toString) {
+      lockManager.executeSequentially(storageId.toString) {
+        if (versionCache.contains((category, storageId))) {
+          Future.successful(versionCache((category, storageId)).asInstanceOf[Seq[OffsetDateTime]])
+        } else {
+          async { // linter:ignore UnnecessaryElseBranch
+            val children = await(store.versions(id).toMat(Sink.seq)(Keep.right).run())
+            versionCache((category, storageId)) = children
+            children
+          }
+        }
+      }
+    }
+    Source.fromFuture(versionsFuture).mapConcat(identity)
+  }
 
   override def deleteVersion[Id, V](
     k: Id,
     version: OffsetDateTime)(implicit ir: IdResolver[Id, V, Category, K]): Future[Done] =
-    store.deleteVersion(k, version)
+    deleteCurrentOrAll(k, () => store.deleteVersion(k, version))
 
   override def toString: String = s"LazyCachingPersistenceStore($store)"
 }
+
+object LazyCachingPersistenceStore {
+
+  def apply[K, Category, Serialized](
+    store: BasePersistenceStore[K, Category, Serialized])(implicit
+    mat: Materializer,
+    ctx: ExecutionContext) = new LazyCachingPersistenceStore(store, VersionedValueCacheConfig.Default)
+
+  case class VersionedValueCacheConfig(
+    val maxEntries: Int = MAX_VERSIONED_VALUE_CACHE_SIZE,
+    val purgeCount: Int = PURGE_COUNT_FROM_VERSIONED_VALUE_CACHE,
+    val pRemove: Double = P_REMOVE_FROM_VERSIONED_VALUE_CACHE
+  )
+
+  object VersionedValueCacheConfig {
+    val Default = VersionedValueCacheConfig()
+  }
+
+  /**
+    * max number of entries allowed in versionedValueCache before entries are purged
+    */
+  val MAX_VERSIONED_VALUE_CACHE_SIZE = 10000
```

Comment:
MaxVersionedValueCacheSize - Please use scala style constants, not java style. This is also completely consistent with StorageConfig and all of this stuff should be _there_ not here.